File: 1.2-threads/csharp/ex01/Controller/SomaController.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ex01.Tasks;

namespace ex01.Controller;

public class SomaController
{
    public async Task<int> CalcularSomaTotalAsync(List<List<int>> partes)
    {
        if (partes == null || partes.Count == 0)
        {
            return 0;
        }

        List<Task<int>> tarefas = new();

        for (int i = 0; i < partes.Count; i++)
        {
            int numeroParte = i + 1;
            SomaTask task = new(partes[i], numeroParte);
            tarefas.Add(Task.Run(() => task.Executar()));
        }

        int[] resultados = await Task.WhenAll(tarefas);

        return resultados.Sum();
    }
}

File: 1.2-threads/java/ex01/src/controller/SomaController.java
package controller;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import thread.SomaTask;

/**
 * Classe responsavel por coordenar a execucao das tarefas de soma
 * e agregar os resultados parciais.
 */
public class SomaController {

    /**
     * recebe as sublistas de numeros, cria e executa uma SomaTask para cada sublista,
     * aguarda todas terminarem e retorna a soma total
     * @param partes lista contendo sublistas de numeros
     * @return a soma total de todos os numeros
     */
    public int calcularSomaTotal(List<List<Integer>> partes) {

        if (partes == null || partes.isEmpty()) {
            return 0;
        }

        int somaTotal = 0;
    
        ExecutorService executor = Executors.newFixedThreadPool(partes.size());

        try {
            List<Future<Integer>> futuros = new ArrayList<>();

            // dispara uma SomaTask para cada sublista
            for (List<Integer> parte : partes) {
                futuros.add(executor.submit(new SomaTask(parte)));
            }

            // coleta cada soma parcial
            for (Future<Integer> futuro : futuros) {
                try {
                    somaTotal += futuro.get();

                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    System.err.println("Thread interrompida durante a execução da tarefa: " + e.getMessage());

                } catch (ExecutionException e) {
                    System.err.println("Erro na execução da tarefa de soma: " + e.getCause());
                }
            }
        } finally {
            executor.shutdown();
        }

        return somaTotal;
    }
}

